bigtable/metrics.go

```diff
@@ -0,0 +1,487 @@
+			// User provided meter provider
+			tracerFactory.builtinEnabled = true
+			meterProvider = v.MeterProvider
+		case NoopMetricsProvider:
+			tracerFactory.builtinEnabled = false
+			return tracerFactory, nil
+		default:
+			tracerFactory.builtinEnabled = false
+			return tracerFactory, errors.New("Unknown MetricsProvider type")
+		}
+	}
+
+	// Create meter and instruments
+	meter := meterProvider.Meter(builtInMetricsMeterName, metric.WithInstrumentationVersion(internal.Version))
+	err = tracerFactory.createInstruments(meter)
+	return tracerFactory, err
+}
+
+func (tf *builtinMetricsTracerFactory) createInstruments(meter metric.Meter) error {
+	var err error
+
+	// Create operation_latencies
+	tf.operationLatencies, err = meter.Float64Histogram(
+		metricNameOperationLatencies,
+		metric.WithDescription("Total time until final operation success or failure, including retries and backoff."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create attempt_latencies
+	tf.attemptLatencies, err = meter.Float64Histogram(
+		metricNameAttemptLatencies,
+		metric.WithDescription("Client observed latency per RPC attempt."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create server_latencies
+	tf.serverLatencies, err = meter.Float64Histogram(
+		metricNameServerLatencies,
+		metric.WithDescription("The latency measured from the moment that the RPC entered the Google data center until the RPC was completed."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create retry_count
+	tf.retryCount, err = meter.Int64Counter(
+		metricNameRetryCount,
+		metric.WithDescription("The number of additional RPCs sent after the initial attempt."),
+	)
+	return err
+}
+
+// builtinMetricsTracer is created one per function call
+// It is used to store metric instruments, attribute values
+// and other data required to obtain and record them
+type builtinMetricsTracer struct {
+	ctx            context.Context
+	builtInEnabled bool
+	opStartTime    time.Time
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	instrumentOperationLatencies metric.Float64Histogram
+	instrumentServerLatencies    metric.Float64Histogram
+	instrumentAttemptLatencies   metric.Float64Histogram
+	instrumentRetryCount         metric.Int64Counter
+
+	tableName    string
+	appProfileID string
+	method       string
+	isStreaming  bool
+
+	// gRPC status code
+	status string
+
+	// Contains the header response metadata which is used to extract cluster and zone
+	headerMD *metadata.MD
+
+	// Contains the trailer response metadata which is used to extract cluster and zone
+	trailerMD *metadata.MD
+
+	attemptCount int64
```

Comment:
There is some ambiguity here as to who owns each attribute. ie it seems like the client populates everything except the attemptCount. I think it would be better to make this struct own its own state and have the client invoke a method to update internal state.

Reply:
Updated